File: src/main/java/io/github/hdzitao/editstarters/cache/InitializrCache.java
package io.github.hdzitao.editstarters.cache;

import com.intellij.openapi.components.PersistentStateComponent;
import com.intellij.openapi.components.RoamingType;
import com.intellij.openapi.components.State;
import com.intellij.openapi.components.Storage;
import com.intellij.openapi.project.Project;
import com.intellij.util.xmlb.annotations.OptionTag;
import io.github.hdzitao.editstarters.ohub.GitHub;
import io.github.hdzitao.editstarters.springboot.SpringBoot;
import lombok.Getter;
import lombok.Setter;
import org.apache.commons.lang3.StringUtils;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.util.Objects;

/**
 * 缓存
 *
 * @version 3.2.0
 */
@State(name = "editstarters-initializr-cache",
        storages = @Storage(value = "editstarters/initializr-cache.xml", roamingType = RoamingType.DISABLED))
public class InitializrCache implements PersistentStateComponent<InitializrCache.State> {

    public static InitializrCache getInstance(Project project) {
        return project.getService(InitializrCache.class);
    }

    /**
     * 缓存bean
     */
    @Getter
    @Setter
    public static class State {
        @OptionTag(converter = SpringBootConverter.class)
        private SpringBoot springBoot;
        private String url;
        private String oHub;
        private String version;
        private long updateTime;
    }

    public static class SpringBootConverter extends JSONConverter<SpringBoot> {
    }

    private State state;

    /**
     * 缓存是否有效
     */
    public boolean enable() {
        return state != null && StringUtils.isNoneBlank(state.url);
    }

    /**
     * 初始化
     */
    public void initialize() {
        if (!enable()) {
            state = new State();
        }
    }

    /**
     * 获取缓存
     */
    public SpringBoot getSpringBoot(String url, String version) {
        // 检查缓存
        if (enable()
                && Objects.equals(url, state.url)
                && Objects.equals(version, state.version)
                && state.springBoot != null) {
            return state.springBoot;
        }

        return null;
    }

    /**
     * 更新缓存
     */
    public void putSpringBoot(String url, String version, SpringBoot project) {
        state.url = url;
        state.version = version;
        state.springBoot = project;
        state.updateTime = System.currentTimeMillis();
    }

    /**
     * 获取url
     */
    public String getUrl() {
        if (enable() && StringUtils.isNotEmpty(state.url)) {
            return state.url;
        }

        return "https://start.spring.io/";
    }

    /**
     * 获取更新时间
     */
    public long getUpdateTime() {
        if (enable()) {
            return state.updateTime;
        }

        return 0L;
    }

    /**
     * 获取ohub
     */
    public String getOHubName() {
        if (enable() && StringUtils.isNotEmpty(state.oHub)) {
            return state.oHub;
        }

        return GitHub.NAME;
    }

    /**
     * 缓存oHub
     */
    public void putOHubName(String oHubName) {
        state.oHub = oHubName;
    }

    @Nullable
    @Override
    public InitializrCache.State getState() {
        return state;
    }

    @Override
    public void loadState(@NotNull InitializrCache.State state) {
        this.state = state;
    }

}

File: src/main/kotlin/io/github/hdzitao/editstarters/cache/InitializrCache.kt
package io.github.hdzitao.editstarters.cache

import com.intellij.openapi.components.*
import com.intellij.openapi.project.Project
import com.intellij.util.xmlb.annotations.OptionTag
import io.github.hdzitao.editstarters.springboot.SpringBoot

/**
 * 缓存
 *
 * @version 3.2.0
 */
@Service(Service.Level.PROJECT)
@State(
    name = "editstarters-initializr-cache",
    storages = [Storage(value = "editstarters/initializr-cache.xml", roamingType = RoamingType.DISABLED)]
)
class InitializrCache : PersistentStateComponent<InitializrCache.State?> {
    /**
     * 缓存bean
     */
    class State {
        @OptionTag(converter = SpringBootConverter::class)
        var springBoot: SpringBoot? = null
        var url: String? = null
        var oHub: String? = null
        var version: String? = null
        var updateTime: Long = 0
    }

    class SpringBootConverter : JSONConverter<SpringBoot>()

    private var state: State? = null

    /**
     * 获取缓存
     */
    fun getSpringBoot(url: String, version: String): SpringBoot? =
        if (state?.url == url && version == state?.version && state?.springBoot != null) {
            state?.springBoot
        } else {
            null
        }

    /**
     * 更新缓存
     */
    fun putSpringBoot(url: String, version: String, project: SpringBoot) {
        state?.url = url
        state?.version = version
        state?.springBoot = project
        state?.updateTime = System.currentTimeMillis()
    }

    val url: String?
        get() = state?.url

    val updateTime: Long?
        get() = state?.updateTime


    val oHubName: String?
        get() = state?.oHub


    /**
     * 缓存oHub
     */
    fun putOHubName(oHubName: String?) {
        state!!.oHub = oHubName
    }

    override fun getState(): State? {
        return state
    }

    override fun loadState(state: State) {
        this.state = state
    }

    companion object {
        @JvmStatic
        fun getInstance(project: Project): InitializrCache {
            return project.getService(InitializrCache::class.java)
        }
    }
}
